File: packages/python/protocol/starfish_protocol/crypto.py
"""Shared cryptographic primitives for the Starfish sync protocol."""


from typing import Any, Protocol, runtime_checkable

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

IV_BYTES = 12
ENCRYPTED_KEY = "_encrypted"


@runtime_checkable
class Encryptor(Protocol):
    """Encrypt/decrypt contract for client-side E2E encryption.

    Defined as a structural :class:`typing.Protocol` (not an ABC) because
    implementations such as the keyring's ``KeyringEncryptor`` conform by
    shape rather than by subclassing. Lives in the shared protocol layer so
    the client (``SyncManager``) and the implementations can both reference
    the contract without a package dependency cycle.
    """

    def encrypt(self, data: dict[str, Any]) -> dict[str, Any]: ...

    def decrypt(self, wrapper: dict[str, Any]) -> dict[str, Any]: ...


def derive_key(secret: str, salt: str, info: bytes) -> bytes:
    """Derive a 256-bit AES key from a secret and salt using HKDF(SHA-256).

    Mirrors the TypeScript :func:`deriveKey`.
    """
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt.encode("utf-8"),
        info=info,
    )
    return hkdf.derive(secret.encode("utf-8"))

File: packages/ts/protocol/src/crypto.ts
import { getCrypto } from "./platform.js"

const ALGO = "AES-GCM"
export const IV_BYTES = 12
export const ENCRYPTED_KEY = "_encrypted"

/**
 * Encrypt/decrypt contract for client-side E2E encryption.
 *
 * Lives in the shared protocol layer so the client (`SyncManager`, which
 * consumes an `Encryptor` via its `encryptor` option) and the encryptor
 * implementations (e.g. the keyring's `createKeyringEncryptor`) can both
 * reference the contract without a workspace dependency cycle.
 */
export interface Encryptor {
  encrypt(data: Record<string, unknown>): Promise<Record<string, unknown>>
  decrypt(wrapper: Record<string, unknown>): Promise<Record<string, unknown>>
}

/**
 * Derive the raw 32-byte AES-256 key via HKDF-SHA256 over UTF-8(`secret`) with
 * UTF-8(`salt`) / UTF-8(`info`).
 *
 * Exposed for cross-language conformance testing — the shared `crypto.json`
 * vector anchors these exact bytes, and the Python `derive_key` must produce
 * the same. Production code should use {@link deriveKey}, which imports these
 * bytes as a non-extractable AES-GCM key.
 */
export async function deriveAesKeyBytes(
  secret: string,
  salt: string,
  info: string,
): Promise<Uint8Array> {
  const enc = new TextEncoder()
  const c = getCrypto()
  const keyMaterial = await c.subtle.importKey(
    "raw",
    enc.encode(secret),
    "HKDF",
    false,
    ["deriveBits"],
  )
  const bits = await c.subtle.deriveBits(
    {
      name: "HKDF",
      hash: "SHA-256",
      salt: enc.encode(salt),
      info: enc.encode(info),
    },
    keyMaterial,
    256,
  )
  return new Uint8Array(bits)
}

export async function deriveKey(secret: string, salt: string, info: string): Promise<CryptoKey> {
  const keyBytes = await deriveAesKeyBytes(secret, salt, info)
  return getCrypto().subtle.importKey(
    "raw",
    keyBytes as BufferSource,
    { name: ALGO },
    false,
    ["encrypt", "decrypt"],
  )
}
